bpo-30767 - handle empty exception frame and prevent extraneous output

tb3088 opened the pull request:
While it's not common-place when handling an exception, it's possible for an (custom) exception handler to not find a frame after walking the call stack (eg. function _log() line 1445, function handleError() line 920).

When faced with an exc_info of (None, None, None)  an extra line of the form:
`NoneObject None`
being added to the exception message during formatting step.

I also removed the incomplete parameter sanity check and invoked the already defined one.

<!-- issue-number: bpo-30767 -->
https://bugs.python.org/issue30767
<!-- /issue-number -->

tb3088 committed:
handle empty exception frame and prevent extraneous output

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

tb3088 committed:
protect against wrong data type if "raiseExceptions" has been set False

tb3088 committed:
actually enforce valid values or return NOTSET

tb3088 committed:
fix logic to only allow non-empty Tuples thru

tb3088 committed:
checking the 1st element in the Tuple is sufficient

tb3088 committed:
change order of evaluation for level to not except too soon and tolerate arbitrary integers

tb3088 committed:
test harness expects TypeError exception

tb3088 committed:
revert gratuitous change to Exception message

tb3088 committed:
a String that can be cast to an Integer is no longer an error

serhiy-storchaka commented on Lib/test/test_logging.py:
```diff
@@ -4023,11 +4023,11 @@ def test_exception(self):
     def test_log_invalid_level_with_raise(self):
         with support.swap_attr(logging, 'raiseExceptions', True):
-            self.assertRaises(TypeError, self.logger.log, '10', 'test message')
+            self.assertRaises(TypeError, self.logger.log, 'xx', 'test message')
```
Don't change existing tests. Just add new tests.

tb3088 replied:
'10' == 10 so I would need to (re)move this test. What we should be testing is if it's an unknown string, not an integer in hiding.

serhiy-storchaka replied:
No, in Python '10' != 10.

If there is a special test for '10', thus it is rejected for purpose. Until the design decision is not changed, the test shouldn't be changed.

tb3088 replied:
as far as logging levels is concerned '10' == 10. Yes I'm aware that Int != Str. It pointlessly pedantic to throw an exception for '10' when it is trivially re-cast as an Int. It's not an error worth forcing the user to go fix. "Liberal in what you accept. Strict in what you emit".

serhiy-storchaka replied:
Currently this is considered as an error. There are purposed tests for this. If you want to change this and have use cases, please open a special issue on the tracker.

tb3088 committed:
When a non-recognized 'level' is used, just emit the string form.
Changing what is a single field into two and adding the gratuitous
'Level ' is what directly caused issue 29220.

serhiy-storchaka commented on Lib/logging/__init__.py:
```diff
@@ -185,14 +185,29 @@ def currentframe():
+        #FIXME - test harness injects '+1',  so tolerating 
+        # arbitrary integers is expected behavior. Why?
+        #    raise ValueError
+            rv = int(level)
```
This makes `'10'` and `12.3` be accepted now.

tb3088 replied:
if it can be reduced to an integer than it's perfectly "valid". Otherwise we should strictly check for defined levels. Which apparently is not desired.

serhiy-storchaka replied:
This should be discussed on the bug tracker. If this is desired, this needs to be documented.

serhiy-storchaka commented on Lib/logging/__init__.py:
```diff
@@ -185,14 +185,29 @@ def currentframe():
+        if raiseExceptions:
+            # test harness (../test/test_logging) expects 'TypeError'
+            raise TypeError("Level not an integer or a valid string: %r" % level) from err
+    except Exception:
```
This silences exceptions like MemoryError or BytesWarning.

tb3088 replied:
Are those exceptions relevant to this function? We care if it's a known string or an Integer. The original had no exception handling and only trapped TypeError anyhow. 

But if 'raise' is more correct, happy to fix. If so, do we honor 'if raiseExceptions'? Also the test harness will need to be updated to tolerate more than just a TypeError being raised.

serhiy-storchaka replied:
Yes, these exception can be raised in this function. If silencing them is desired, this should be explained and tested.

serhiy-storchaka commented on Lib/logging/__init__.py:
```diff
@@ -569,7 +584,9 @@ def format(self, record):
             record.asctime = self.formatTime(record, self.datefmt)
         s = self.formatMessage(record)
-        if record.exc_info:
+        if (isinstance(record.exc_info, tuple) and record.exc_info[0]):
```
What can it be if not a tuple?

tb3088 replied:
Boolean. format() tries to deference it as an item which obviously doesn't play very nicely. An empty tuple (None, None, None) isn't handled by formatter and thus emits junk. Perhaps the better fix is to add None detection to the formatter to short-circuit any output?

serhiy-storchaka replied:
According to the documentation, exc_info is an exception tuple with the current exception information, or None if no exception information is available. Hence `isinstance(record.exc_info, tuple)`, `record.exc_info is not None` and just `record.exc_info` should be equivalent, but the latter form is the shortest. If needed a special case for the tuple `(None, None, None)` (this is not an *empty tuple*), it should be tested.

tb3088 replied:
The documentation also says you can call log() with exc_info='a True-like value'. So invoking log() with exc_info=True  is also acceptable practice and examples abound. Furthermore sys.exc_info() hasn't necessarily been called yet to generate the tuple. When exc_info() doesn't find a stack frame it returns (None, None, None) which you are correct is, not None. Point is Boolean does get past in and the code just assumes a Tuple will ALWAYS be the data type and blindly de-references it. The formatter also blindly assumes a valid stack frame was found and that the Tuple can't be None^3.

vsajip reviewed (commented):
I am not keen to merge this PR as bpo-30767 does not appear to be a real issue, IMO (see the issue tracker for comment). An example script showing the alleged problem should be posted on the issue.
